<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>data_integration</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="Work_Report_Yilin_8.16_files/libs/clipboard/clipboard.min.js"></script>
<script src="Work_Report_Yilin_8.16_files/libs/quarto-html/quarto.js"></script>
<script src="Work_Report_Yilin_8.16_files/libs/quarto-html/popper.min.js"></script>
<script src="Work_Report_Yilin_8.16_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="Work_Report_Yilin_8.16_files/libs/quarto-html/anchor.min.js"></script>
<link href="Work_Report_Yilin_8.16_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="Work_Report_Yilin_8.16_files/libs/quarto-html/quarto-syntax-highlighting-2f5df379a58b258e96c21c0638c20c03.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="Work_Report_Yilin_8.16_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="Work_Report_Yilin_8.16_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="Work_Report_Yilin_8.16_files/libs/bootstrap/bootstrap-1bc8a17f135ab3d594c857e9f48e611b.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">data_integration</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="batch-effect-correction-pipeline" class="level1">
<h1>Batch effect correction pipeline</h1>
<section id="pre-processing" class="level2">
<h2 class="anchored" data-anchor-id="pre-processing">pre-processing</h2>
<p>There are necessary steps before correction– filtering, normalization, identifying highly variable genes, scaling the expression matrix.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Matrix)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(sparseMatrixStats)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>load dataset</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>read_GSE115978 <span class="ot">&lt;-</span> <span class="cf">function</span>(data_path) {</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Reading GSE115978 dataset using fread...</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>  counts_path <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_path, <span class="st">"GSE115978"</span>, <span class="st">"GSE115978_counts.csv.gz"</span>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>  anno_path <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_path, <span class="st">"GSE115978"</span>, <span class="st">"GSE115978_cell.annotations.csv.gz"</span>)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>  counts_dt <span class="ot">&lt;-</span> <span class="fu">fread</span>(counts_path)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>  gene_names <span class="ot">&lt;-</span> counts_dt[[<span class="dv">1</span>]]</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>  counts_mat <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(counts_dt[, <span class="sc">-</span><span class="dv">1</span>, <span class="at">with =</span> <span class="cn">FALSE</span>])</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames</span>(counts_mat) <span class="ot">&lt;-</span> gene_names</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>  anno_df <span class="ot">&lt;-</span> <span class="fu">fread</span>(anno_path)</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>  anno_df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(anno_df)</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames</span>(anno_df) <span class="ot">&lt;-</span> anno_df<span class="sc">$</span>cells</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>  common_cells <span class="ot">&lt;-</span> <span class="fu">intersect</span>(<span class="fu">colnames</span>(counts_mat), anno_df<span class="sc">$</span>cells)</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>  counts_mat <span class="ot">&lt;-</span> counts_mat[, common_cells, drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>  anno_df <span class="ot">&lt;-</span> anno_df[common_cells, , drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>  seurat_obj <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(</span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">counts =</span> counts_mat,</span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>    <span class="at">meta.data =</span> anno_df,</span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>    <span class="at">project =</span> <span class="st">"GSE115978"</span></span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a>  seurat_obj<span class="sc">$</span>batch <span class="ot">&lt;-</span> <span class="st">"GSE115978"</span></span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a>  seurat_obj<span class="sc">$</span>study <span class="ot">&lt;-</span> <span class="st">"GSE115978"</span></span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"GSE115978:"</span>, <span class="fu">ncol</span>(seurat_obj), <span class="st">"cells,"</span>, <span class="fu">nrow</span>(seurat_obj), <span class="st">"genes</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(seurat_obj)</span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true" tabindex="-1"></a>data_path<span class="ot">&lt;-</span> <span class="st">"C:/Users/Administrator/Downloads"</span></span>
<span id="cb2-33"><a href="#cb2-33" aria-hidden="true" tabindex="-1"></a>reads_GSE115978<span class="ot">&lt;-</span><span class="fu">read_GSE115978</span>(data_path)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Reading GSE115978 dataset using fread...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Feature names cannot have underscores ('_'), replacing with dashes
('-')</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Data is of class data.frame. Coercing to dgCMatrix.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>GSE115978: 7186 cells, 23686 genes</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(reads_GSE115978<span class="sc">$</span>nFeature_RNA,<span class="dv">3</span>)  <span class="co"># colSums(counts&gt;0) non-zero entries</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           cy78_CD45_neg_1_B04_S496_comb 
                                    8094 
cy79_p4_CD45_neg_PDL1_neg_E11_S1115_comb 
                                    2022 
                    CY88_5_B10_S694_comb 
                                    5141 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(reads_GSE115978<span class="sc">$</span>nCount_RNA,<span class="dv">3</span>)  <span class="co"># colSums(counts) total number of UMI counts per cell</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           cy78_CD45_neg_1_B04_S496_comb 
                                  357919 
cy79_p4_CD45_neg_PDL1_neg_E11_S1115_comb 
                                    5727 
                    CY88_5_B10_S694_comb 
                                  139218 </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(reads_GSE115978<span class="sc">$</span>nCount_RNA,</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">breaks =</span> <span class="dv">100</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Work_Report_Yilin_8.16_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Calculate for each cell its total number of reads. It is zero-inflated because of dropout event,commonly seen in sc RNA analysis. AND this histogram shows a strong right-skewed distribution. Normally, we do log-transform to reduce skewness.</p>
<pre><code>::: {.cell}

```{.r .cell-code}
hist(log10(reads_GSE115978$nCount_RNA),breaks=100)</code></pre>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Work_Report_Yilin_8.16_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<p>:::</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co">#GSE115978</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>seurat_115978 <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(<span class="at">object =</span> reads_GSE115978, <span class="at">normalization.method =</span> <span class="st">"LogNormalize"</span>, <span class="at">scale.factor =</span> <span class="dv">10000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Normalizing layer: counts</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>count_matrix<span class="ot">&lt;-</span><span class="fu">GetAssayData</span>(reads_GSE115978, <span class="at">slot =</span> <span class="st">"counts"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The `slot` argument of `GetAssayData()` is deprecated as of SeuratObject 5.0.0.
ℹ Please use the `layer` argument instead.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>frac <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">t</span>(count_matrix) <span class="sc">/</span> reads_GSE115978<span class="sc">$</span>nCount_RNA)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span> (<span class="fu">rowMeans</span>(frac),<span class="fu">rowVars</span>(frac)<span class="sc">/</span><span class="fu">rowMeans</span>(frac),<span class="at">log=</span><span class="st">"xy"</span>,<span class="at">cex=</span>.<span class="dv">2</span>,<span class="at">col=</span><span class="fu">adjustcolor</span>(<span class="st">"black"</span>,<span class="at">alpha=</span>.<span class="dv">2</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in xy.coords(x, y, xlabel, ylabel, log): 750 x values &lt;= 0 omitted from
logarithmic plot</code></pre>
</div>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>ntilde<span class="ot">&lt;-</span><span class="dv">1</span><span class="sc">/</span><span class="fu">mean</span>(<span class="dv">1</span><span class="sc">/</span> reads_GSE115978<span class="sc">$</span>nCount_RNA)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h=</span><span class="dv">1</span><span class="sc">/</span>ntilde,<span class="at">col=</span><span class="st">"purple"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Work_Report_Yilin_8.16_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>seurat_115978 <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(<span class="at">object =</span> seurat_115978, <span class="at">selection.method =</span> <span class="st">"vst"</span>,</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">nfeatures =</span> <span class="dv">2000</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">VariableFeatures</span>(seurat_115978))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "COL3A1" "DCN"    "COL1A1" "COL1A2" "SPP1"   "ELK2AP"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>seurat_115978 <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>  seurat_115978, <span class="at">features =</span> <span class="fu">VariableFeatures</span>(seurat_115978)</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Centering and scaling data matrix</code></pre>
</div>
</div>
</section>
<section id="dimension-reduction" class="level2">
<h2 class="anchored" data-anchor-id="dimension-reduction">Dimension Reduction</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co">#mat &lt;- t(GetAssayData(seurat_115978, slot = "scale.data"))</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="co">#pca &lt;- irlba::prcomp_irlba( mat, n=20, center=TRUE, scale.=TRUE )</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="co">#plot(pca$x[,1],pca$x[,2],cex=.1,col="#00000030",asp=1,ylim=c(-20,20))</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>seurat_115978 <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>  seurat_115978,</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">features =</span> <span class="fu">VariableFeatures</span>(seurat_115978),</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">npcs =</span> <span class="dv">20</span>,</span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">verbose =</span> <span class="cn">FALSE</span></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(seurat_115978, <span class="at">reduction =</span> <span class="st">"pca"</span>, <span class="at">dims =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="at">group.by =</span> <span class="st">"samples"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Work_Report_Yilin_8.16_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co">#uwot::umap(pca$x)-&gt;umap</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="co">#plot(umap,cex=.1,col="#00000030",asp=1)</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>seurat_115978 <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>  seurat_115978, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>,</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">umap.method =</span> <span class="st">"uwot"</span>, <span class="at">n.neighbors =</span> <span class="dv">15</span>, <span class="at">min.dist =</span> <span class="fl">0.3</span>, <span class="at">seed.use =</span> <span class="dv">42</span></span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric
To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'
This message will be shown once per session</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>09:31:57 UMAP embedding parameters a = 0.9922 b = 1.112</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>09:31:57 Read 7186 rows and found 20 numeric columns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>09:31:57 Using Annoy for neighbor search, n_neighbors = 15</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>09:31:57 Building Annoy index with metric = cosine, n_trees = 50</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>0%   10   20   30   40   50   60   70   80   90   100%</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>[----|----|----|----|----|----|----|----|----|----|</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>**************************************************|
09:31:58 Writing NN index file to temp file C:\Users\Administrator\AppData\Local\Temp\RtmpIx5CN7\file51082436c86
09:31:58 Searching Annoy index using 1 thread, search_k = 1500
09:32:00 Annoy recall = 100%
09:32:00 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 15
09:32:01 Found 3 connected components, falling back to 'spca' initialization with init_sdev = 1
09:32:01 Using 'irlba' for PCA
09:32:01 PCA: 2 components explained 34.92% variance
09:32:01 Scaling init to sdev = 1
09:32:01 Commencing optimization for 500 epochs, with 159498 positive edges
09:32:01 Using rng type: pcg
09:32:21 Optimization finished</code></pre>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(seurat_115978, <span class="at">reduction =</span> <span class="st">"umap"</span>, <span class="at">group.by =</span> <span class="st">"samples"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Work_Report_Yilin_8.16_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(seurat_115978, <span class="at">reduction =</span> <span class="st">"umap"</span>, <span class="at">group.by =</span> <span class="st">"treatment.group"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Work_Report_Yilin_8.16_files/figure-html/unnamed-chunk-12-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The UMAP plot colored by treatment group shows good mixing— no obvious treatment-related batch effect.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(seurat_115978, <span class="at">reduction =</span> <span class="st">"umap"</span>, <span class="at">group.by =</span> <span class="st">"cell.types"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Work_Report_Yilin_8.16_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Major lineages like T and B cells are separated, but T-cell subtypes (T.cell and T.CD4 and T.CD8 overlap) are hard to tell, which is biologically sensible. This plot also shows that the differences within each lineage are more subtle.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(seurat_115978, <span class="at">features=</span><span class="fu">c</span>(<span class="st">"CD3D"</span>,<span class="st">"IL7R"</span>,<span class="st">"CD8A"</span>,<span class="st">"NKG7"</span>,<span class="st">"MS4A1"</span>,<span class="st">"LYZ"</span>,<span class="st">"EPCAM"</span>,<span class="st">"PECAM1"</span>,<span class="st">"COL1A1"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The `slot` argument of `FetchData()` is deprecated as of SeuratObject 5.0.0.
ℹ Please use the `layer` argument instead.
ℹ The deprecated feature was likely used in the Seurat package.
  Please report the issue at &lt;https://github.com/satijalab/seurat/issues&gt;.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Work_Report_Yilin_8.16_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>canonical lineage markers.</p>
<p>#Unsupervised clustering, graph-based community detection</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>seurat_obj <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(seurat_115978, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="at">k.param =</span> <span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing nearest neighbor graph</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing SNN</code></pre>
</div>
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="do">##normal matrix</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>adjm <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(seurat_obj<span class="sc">@</span>graphs<span class="sc">$</span>RNA_nn) </span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(adjm[, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                                cy78_CD45_neg_1_B04_S496_comb
cy78_CD45_neg_1_B04_S496_comb                                               1
cy79_p4_CD45_neg_PDL1_neg_E11_S1115_comb                                    0
CY88_5_B10_S694_comb                                                        0
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_F07_S67_comb                             0
cy78_CD45_neg_3_H06_S762_comb                                               0
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_G01_S73_comb                             0
                                                cy79_p4_CD45_neg_PDL1_neg_E11_S1115_comb
cy78_CD45_neg_1_B04_S496_comb                                                          0
cy79_p4_CD45_neg_PDL1_neg_E11_S1115_comb                                               1
CY88_5_B10_S694_comb                                                                   0
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_F07_S67_comb                                        0
cy78_CD45_neg_3_H06_S762_comb                                                          0
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_G01_S73_comb                                        0
                                                CY88_5_B10_S694_comb
cy78_CD45_neg_1_B04_S496_comb                                      0
cy79_p4_CD45_neg_PDL1_neg_E11_S1115_comb                           0
CY88_5_B10_S694_comb                                               1
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_F07_S67_comb                    0
cy78_CD45_neg_3_H06_S762_comb                                      0
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_G01_S73_comb                    0
                                                cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_F07_S67_comb
cy78_CD45_neg_1_B04_S496_comb                                                                 0
cy79_p4_CD45_neg_PDL1_neg_E11_S1115_comb                                                      0
CY88_5_B10_S694_comb                                                                          0
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_F07_S67_comb                                               1
cy78_CD45_neg_3_H06_S762_comb                                                                 0
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_G01_S73_comb                                               0
                                                cy78_CD45_neg_3_H06_S762_comb
cy78_CD45_neg_1_B04_S496_comb                                               0
cy79_p4_CD45_neg_PDL1_neg_E11_S1115_comb                                    0
CY88_5_B10_S694_comb                                                        0
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_F07_S67_comb                             0
cy78_CD45_neg_3_H06_S762_comb                                               1
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_G01_S73_comb                             0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="do">##sparse matrix</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>adjm_sparse <span class="ot">&lt;-</span> seurat_obj<span class="sc">@</span>graphs<span class="sc">$</span>RNA_nn</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(adjm_sparse[, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>6 x 5 sparse Matrix of class "dgCMatrix"
                                                cy78_CD45_neg_1_B04_S496_comb
cy78_CD45_neg_1_B04_S496_comb                                               1
cy79_p4_CD45_neg_PDL1_neg_E11_S1115_comb                                    .
CY88_5_B10_S694_comb                                                        .
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_F07_S67_comb                             .
cy78_CD45_neg_3_H06_S762_comb                                               .
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_G01_S73_comb                             .
                                                cy79_p4_CD45_neg_PDL1_neg_E11_S1115_comb
cy78_CD45_neg_1_B04_S496_comb                                                          .
cy79_p4_CD45_neg_PDL1_neg_E11_S1115_comb                                               1
CY88_5_B10_S694_comb                                                                   .
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_F07_S67_comb                                        .
cy78_CD45_neg_3_H06_S762_comb                                                          .
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_G01_S73_comb                                        .
                                                CY88_5_B10_S694_comb
cy78_CD45_neg_1_B04_S496_comb                                      .
cy79_p4_CD45_neg_PDL1_neg_E11_S1115_comb                           .
CY88_5_B10_S694_comb                                               1
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_F07_S67_comb                    .
cy78_CD45_neg_3_H06_S762_comb                                      .
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_G01_S73_comb                    .
                                                cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_F07_S67_comb
cy78_CD45_neg_1_B04_S496_comb                                                                 .
cy79_p4_CD45_neg_PDL1_neg_E11_S1115_comb                                                      .
CY88_5_B10_S694_comb                                                                          .
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_F07_S67_comb                                               1
cy78_CD45_neg_3_H06_S762_comb                                                                 .
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_G01_S73_comb                                               .
                                                cy78_CD45_neg_3_H06_S762_comb
cy78_CD45_neg_1_B04_S496_comb                                               .
cy79_p4_CD45_neg_PDL1_neg_E11_S1115_comb                                    .
CY88_5_B10_S694_comb                                                        .
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_F07_S67_comb                             .
cy78_CD45_neg_3_H06_S762_comb                                               1
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_G01_S73_comb                             .</code></pre>
</div>
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colSums</span>(adjm_sparse) <span class="ot">-&gt;</span> vertex_degrees</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>( vertex_degrees,<span class="at">breaks=</span><span class="dv">30</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Work_Report_Yilin_8.16_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>A few cells have very high degrees– these might be hub nodes cells with very low degrees may be isolated in te data, could be labeled as outliers.</p>
<p>seurat_obj@graphs$RNA_nn is a symmetric adjacency</p>
<p>we check hub cells which</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>mean_deg <span class="ot">&lt;-</span> <span class="fu">mean</span>(vertex_degrees)</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>sd_deg <span class="ot">&lt;-</span> <span class="fu">sd</span>(vertex_degrees)</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>hub_threshold <span class="ot">&lt;-</span> mean_deg <span class="sc">+</span> <span class="dv">3</span> <span class="sc">*</span> sd_deg</span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a>hub_cells <span class="ot">&lt;-</span> <span class="fu">which</span>(vertex_degrees <span class="sc">&gt;</span> hub_threshold)</span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"The number of Hub cells :"</span>, <span class="fu">length</span>(hub_cells), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>The number of Hub cells : 129 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(vertex_degrees, <span class="at">breaks=</span><span class="dv">50</span>, <span class="at">main=</span><span class="st">"Degree distribution with thresholds"</span>, <span class="at">xlab=</span><span class="st">"Degree"</span>)</span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v =</span> hub_threshold, <span class="at">col=</span><span class="st">"red"</span>, <span class="at">lwd=</span><span class="dv">2</span>, <span class="at">lty=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Work_Report_Yilin_8.16_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>undirected graph</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>pca_matrix <span class="ot">&lt;-</span> seurat_obj<span class="sc">@</span>reductions<span class="sc">$</span>pca<span class="sc">@</span>cell.embeddings</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(pca_matrix)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                                     PC_1      PC_2      PC_3
cy78_CD45_neg_1_B04_S496_comb                    7.757568 -4.815407 -7.825449
cy79_p4_CD45_neg_PDL1_neg_E11_S1115_comb         7.637491 -1.780396 -8.320057
CY88_5_B10_S694_comb                             5.092996 -3.291351 -6.988809
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_F07_S67_comb 10.583775 -1.268354 -8.901335
cy78_CD45_neg_3_H06_S762_comb                    9.320724 -3.336769 -7.876569
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_G01_S73_comb 11.137878 -2.044551 -9.909791
                                                      PC_4       PC_5
cy78_CD45_neg_1_B04_S496_comb                   -8.4726691 -2.9296714
cy79_p4_CD45_neg_PDL1_neg_E11_S1115_comb         3.7936347 -1.0521534
CY88_5_B10_S694_comb                             2.6271141 -1.7388752
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_F07_S67_comb  3.0858221 -0.2944948
cy78_CD45_neg_3_H06_S762_comb                    0.1114288 -3.4952191
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_G01_S73_comb  4.1340678 -0.4131963
                                                      PC_6       PC_7
cy78_CD45_neg_1_B04_S496_comb                    7.7260618  4.6757712
cy79_p4_CD45_neg_PDL1_neg_E11_S1115_comb         1.1980493  1.2223539
CY88_5_B10_S694_comb                             3.9080013  3.2654034
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_F07_S67_comb -1.2088186 -1.1015186
cy78_CD45_neg_3_H06_S762_comb                    7.8290884  4.8168332
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_G01_S73_comb -0.2757132  0.4911571
                                                      PC_8       PC_9
cy78_CD45_neg_1_B04_S496_comb                   -6.5844264 10.8176718
cy79_p4_CD45_neg_PDL1_neg_E11_S1115_comb         7.9090228 -0.3352334
CY88_5_B10_S694_comb                            -0.5079443 -2.1169324
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_F07_S67_comb 11.1895731  0.5785861
cy78_CD45_neg_3_H06_S762_comb                   -6.0242500 -0.1855716
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_G01_S73_comb 12.0975556  0.7604812
                                                     PC_10     PC_11      PC_12
cy78_CD45_neg_1_B04_S496_comb                   -2.2210212  2.365666  4.5648192
cy79_p4_CD45_neg_PDL1_neg_E11_S1115_comb         0.4601099 -3.345261 -1.3828755
CY88_5_B10_S694_comb                             1.4811888  2.330527 -0.2079382
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_F07_S67_comb  0.8494759 -3.350415  1.4911046
cy78_CD45_neg_3_H06_S762_comb                    2.4267101  2.421964  4.5126752
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_G01_S73_comb  0.4404555 -3.776621  2.3118006
                                                     PC_13      PC_14
cy78_CD45_neg_1_B04_S496_comb                   -0.8441064 -1.5232913
cy79_p4_CD45_neg_PDL1_neg_E11_S1115_comb        -2.5263964  0.1789084
CY88_5_B10_S694_comb                            -0.3470690 -1.7058970
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_F07_S67_comb -1.6749182  0.3978231
cy78_CD45_neg_3_H06_S762_comb                    3.3313742  0.3038922
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_G01_S73_comb -3.6045924  0.7901843
                                                      PC_15      PC_16
cy78_CD45_neg_1_B04_S496_comb                   -1.08921550  1.4890489
cy79_p4_CD45_neg_PDL1_neg_E11_S1115_comb         0.03415717  0.9445262
CY88_5_B10_S694_comb                            -0.55591907 -0.3232698
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_F07_S67_comb -3.07852643  1.4573855
cy78_CD45_neg_3_H06_S762_comb                    0.96577639 -1.0245448
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_G01_S73_comb -1.75521706  1.2766905
                                                       PC_17       PC_18
cy78_CD45_neg_1_B04_S496_comb                    3.462213121 -0.92565924
cy79_p4_CD45_neg_PDL1_neg_E11_S1115_comb         0.001726959  0.09885860
CY88_5_B10_S694_comb                             0.483203838 -1.01718699
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_F07_S67_comb -2.614309239  0.08287981
cy78_CD45_neg_3_H06_S762_comb                    4.358801472 -3.37882629
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_G01_S73_comb -1.508694965 -0.75657241
                                                     PC_19      PC_20
cy78_CD45_neg_1_B04_S496_comb                   -5.0439151 -0.6163044
cy79_p4_CD45_neg_PDL1_neg_E11_S1115_comb         2.0303923 -0.9019998
CY88_5_B10_S694_comb                            -2.8351141 -0.9434532
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_F07_S67_comb -0.9031245  2.9044848
cy78_CD45_neg_3_H06_S762_comb                   -4.1899826  1.0324337
cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_G01_S73_comb  1.5146982 -0.5052203</code></pre>
</div>
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(FNN)</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>nn <span class="ot">&lt;-</span> <span class="fu">get.knn</span>(pca_matrix, <span class="at">k =</span> <span class="dv">20</span>)<span class="sc">$</span>nn.index</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(nn)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13] [,14]
[1,]  473  480   25 1049   24  447  270  874 1146   399   620   864    35  6702
[2,]  572 1045  786   45  516  253  567 1042  423   644   863  1079  1137   502
[3,]  150  655  898  708  271 2944  618  547  414   897    14   594  1096   380
[4,]  453  546  925  891  995 3310  539 1102  207   692   753   911   904   127
[5,]  965   99 1014 1029  434  415  739  303  936   570   325   796   464   810
[6,]  911  362  349 1057 1033  528 1005  645  714   102    95  1038  1140   800
     [,15] [,16] [,17] [,18] [,19] [,20]
[1,]   926   963  3272   301   191  4753
[2,]    64   234   511   260   222   175
[3,]   855   513  1011   614   136  1106
[4,]   768   388   726    98   645   329
[5,]   638   529   499   278  1036    36
[6,]   925   891    41   218   135   918</code></pre>
</div>
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Matrix)</span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>adjm_directed <span class="ot">&lt;-</span> <span class="fu">sparseMatrix</span>(</span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">i =</span> <span class="fu">rep</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(nn), <span class="fu">ncol</span>(nn)),  </span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">j =</span> <span class="fu">as.vector</span>(nn),              </span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> <span class="dv">1</span>,</span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">dims =</span> <span class="fu">c</span>(<span class="fu">nrow</span>(nn), <span class="fu">nrow</span>(nn))</span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb55-8"><a href="#cb55-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-9"><a href="#cb55-9" aria-hidden="true" tabindex="-1"></a>adjm_undirected <span class="ot">&lt;-</span> adjm_directed <span class="sc">+</span> <span class="fu">t</span>(adjm_directed)  </span>
<span id="cb55-10"><a href="#cb55-10" aria-hidden="true" tabindex="-1"></a>adjm_undirected[adjm_undirected <span class="sc">&gt;</span> <span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="dv">1</span>            </span>
<span id="cb55-11"><a href="#cb55-11" aria-hidden="true" tabindex="-1"></a>vertex_degrees_undirected <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(adjm_undirected)</span>
<span id="cb55-12"><a href="#cb55-12" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(vertex_degrees_undirected,<span class="at">breaks=</span><span class="dv">50</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Work_Report_Yilin_8.16_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The minimum is 20 but a vertex can habe much more neighbors due to non-mutual neighboprhood relations.</p>
<p>Now we use the igraph package to build this graph and perform modularity clustering</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>igraph<span class="sc">::</span><span class="fu">graph_from_adjacency_matrix</span>(adjm_undirected , <span class="at">mode=</span><span class="st">"undirected"</span> ) <span class="ot">-&gt;</span> nn_graph</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>nn_graph</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>IGRAPH 5af7aa2 U--- 7186 107911 -- 
+ edges from 5af7aa2:
 [1] 1--  24 1--  25 1--  35 1-- 191 1-- 270 1-- 301 1-- 399 1-- 447 1-- 473
[10] 1-- 480 1-- 620 1-- 653 1-- 864 1-- 874 1-- 926 1-- 963 1--1049 1--1146
[19] 1--3272 1--4753 1--4850 1--6702 2--  45 2--  64 2-- 143 2-- 175 2-- 200
[28] 2-- 222 2-- 234 2-- 253 2-- 260 2-- 300 2-- 355 2-- 411 2-- 423 2-- 502
[37] 2-- 511 2-- 516 2-- 567 2-- 572 2-- 592 2-- 625 2-- 633 2-- 644 2-- 773
[46] 2-- 786 2-- 863 2--1028 2--1042 2--1045 2--1079 2--1137 2--3327 2--3648
[55] 2--3838 3--  14 3-- 136 3-- 150 3-- 271 3-- 380 3-- 414 3-- 513 3-- 547
[64] 3-- 594 3-- 614 3-- 618 3-- 655 3-- 708 3-- 713 3-- 855 3-- 897 3-- 898
[73] 3--1011 3--1096 3--1106 3--2944 4--  26 4--  98 4-- 127 4-- 207 4-- 329
+ ... omitted several edges</code></pre>
</div>
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>igraph<span class="sc">::</span><span class="fu">cluster_leiden</span>( nn_graph, <span class="at">objective_function=</span><span class="st">"modularity"</span> ) <span class="ot">-&gt;</span> clustering</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(clustering)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Class 'communities'  hidden list of 5
 $ membership : num [1:7186] 1 2 1 2 1 2 2 1 2 3 ...
 $ nb_clusters: num 21
 $ quality    : num 0.824
 $ algorithm  : chr "leiden"
 $ vcount     : num 7186</code></pre>
</div>
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>seurat_115978<span class="sc">$</span>leiden_clusters <span class="ot">&lt;-</span> <span class="fu">factor</span>(clustering<span class="sc">$</span>membership)</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(seurat_115978, <span class="at">reduction =</span> <span class="st">"umap"</span>, <span class="at">group.by =</span> <span class="st">"leiden_clusters"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Work_Report_Yilin_8.16_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>( <span class="fu">all</span>(adjm_undirected<span class="sc">@</span>x<span class="sc">!=</span><span class="dv">0</span>) ) <span class="co">#pass</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tibble)</span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>adjm_triplet <span class="ot">&lt;-</span> <span class="fu">as</span>(adjm_undirected, <span class="st">"dgTMatrix"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>'as(&lt;dgCMatrix&gt;, "dgTMatrix")' is deprecated.
Use 'as(., "TsparseMatrix")' instead.
See help("Deprecated") and help("Matrix-deprecated").</code></pre>
</div>
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>edge_table <span class="ot">&lt;-</span><span class="fu">tibble</span>( <span class="at">from=</span>adjm_triplet<span class="sc">@</span>i<span class="sc">+</span><span class="dv">1</span>, <span class="at">to=</span>adjm_triplet<span class="sc">@</span>j<span class="sc">+</span><span class="dv">1</span> ) <span class="sc">%&gt;%</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a> dplyr<span class="sc">::</span><span class="fu">filter</span>(.data<span class="sc">$</span>from <span class="sc">&lt;</span> .data<span class="sc">$</span>to)</span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(edge_table)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 2
   from    to
  &lt;dbl&gt; &lt;dbl&gt;
1     3    14
2     8    14
3    13    18
4    14    20
5    14    21
6     1    24</code></pre>
</div>
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'dplyr'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:matrixStats':

    count</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:data.table':

    between, first, last</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    filter, lag</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union</code></pre>
</div>
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>edge_table_clustered <span class="ot">&lt;-</span> edge_table <span class="sc">%&gt;%</span></span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>( </span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">from_cluster =</span> clustering<span class="sc">$</span>membership[from],</span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">to_cluster   =</span> clustering<span class="sc">$</span>membership[to]</span>
<span id="cb71-5"><a href="#cb71-5" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb71-6"><a href="#cb71-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-7"><a href="#cb71-7" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(edge_table_clustered)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 4
   from    to from_cluster to_cluster
  &lt;dbl&gt; &lt;dbl&gt;        &lt;dbl&gt;      &lt;dbl&gt;
1     3    14            1          1
2     8    14            1          1
3    13    18            5          5
4    14    20            1          1
5    14    21            1          1
6     1    24            1          1</code></pre>
</div>
</div>
<p>check the quality of the Leiden clustering</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="do">##edges across different clusters</span></span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>( edge_table_clustered<span class="sc">$</span>from_cluster <span class="sc">!=</span> edge_table_clustered<span class="sc">$</span>to_cluster )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 9540</code></pre>
</div>
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="do">##proportion of edges inside clusters</span></span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>( edge_table_clustered<span class="sc">$</span>from_cluster <span class="sc">==</span> edge_table_clustered<span class="sc">$</span>to_cluster ) <span class="ot">-&gt;</span> fraction_inner</span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a>fraction_inner</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.9115905</code></pre>
</div>
</div>
<p>91.2% of edges are internal– a good clustering quality.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>edge_table <span class="sc">%&gt;%</span></span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mutate</span>( <span class="at">to =</span> <span class="fu">sample</span>(to) ) <span class="sc">%&gt;%</span></span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a><span class="fu">mutate</span>( </span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">from_cluster =</span> clustering<span class="sc">$</span>membership[ from ],</span>
<span id="cb77-5"><a href="#cb77-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">to_cluster   =</span> clustering<span class="sc">$</span>membership[ to ] ) <span class="sc">%&gt;%</span></span>
<span id="cb77-6"><a href="#cb77-6" aria-hidden="true" tabindex="-1"></a><span class="fu">summarise</span>( <span class="fu">mean</span>( from_cluster <span class="sc">==</span> to_cluster ) )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 1
  `mean(from_cluster == to_cluster)`
                               &lt;dbl&gt;
1                             0.0868</code></pre>
</div>
</div>
<p>randomly shuffle the “to” column, and calculate the fraction of edges that under this random permutation, still connect nodes wthin the same cluster. If you shuffle many times and take the average, you get a stable estimate of the expected fraction of within-cluster edges under randomness.</p>
<p>This average is called J0.</p>
<p>how much connectivity (total degree) each cluster holds.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tapply</span>( vertex_degrees_undirected, clustering<span class="sc">$</span>membership, sum ) <span class="ot">-&gt;</span> degree_sums</span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a>degree_sums</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    1     2     3     4     5     6     7     8     9    10    11    12    13 
11369 14784  9646  2152  3927  5646   771 11803   999  3683 22730  4976 33175 
   14    15    16    17    18    19    20    21 
24363 21750  2524 25126  2496  3667  7377  2854 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>( degree_sums<span class="sc">^</span><span class="dv">2</span> ) <span class="sc">/</span> <span class="fu">sum</span>( degree_sums )<span class="sc">^</span><span class="dv">2</span> <span class="ot">-&gt;</span> expected_fraction_inner</span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a>expected_fraction_inner</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.08750604</code></pre>
</div>
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a>fraction_inner <span class="sc">-</span> expected_fraction_inner</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.8240845</code></pre>
</div>
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a>clustering<span class="sc">$</span>quality</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.8240876</code></pre>
</div>
</div>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>